charms/katib-controller/src/charm.py

```diff
@@ -111,13 +110,101 @@ def __init__(self, framework):
         )
         self.dashboard_provider = GrafanaDashboardProvider(self)
 
-        for event in [
-            self.on.config_changed,
-            self.on.install,
-            self.on.leader_elected,
-            self.on.upgrade_charm,
-        ]:
-            self.framework.observe(event, self.set_pod_spec)
+        # Charm logic
+        self.charm_reconciler = CharmReconciler(self)
+
+        # Generate self-signed certificates and store them
+        self._gen_certs_if_missing()
+
+        self.leadership_gate = self.charm_reconciler.add(
+            component=LeadershipGateComponent(
+                charm=self,
+                name="leadership-gate",
+            ),
+            depends_on=[],
+        )
+
+        self.kubernetes_resources = self.charm_reconciler.add(
+            component=KubernetesComponent(
+                charm=self,
+                name="kubernetes:auths-webhooks-crds-configmaps",
+                resource_templates=K8S_RESOURCE_FILES,
+                krh_resource_types={
+                    ClusterRole,
+                    ClusterRoleBinding,
+                    CustomResourceDefinition,
+                    ServiceAccount,
+                    MutatingWebhookConfiguration,
+                    ValidatingWebhookConfiguration,
+                    ConfigMap,
+                },
+                krh_labels=create_charm_default_labels(
+                    self.app.name, self.model.name, scope="auths-webhooks-crds-configmaps"
+                ),
+                context_callable=lambda: {
+                    **{
+                        "app_name": self.app.name,
+                        "namespace": self._namespace,
+                        "ca_bundle": b64encode(self._stored.ca.encode("ascii")).decode("utf-8"),
+                        "webhookPort": self.model.config["webhook-port"],
+                    },
+                    **self.get_images(
+                        DEFAULT_IMAGES,
+                        parse_images_config(self.model.config["custom_images"]),
+                    ),
```

Comment:
nice suggestion, on it!

Reply:
yeah what you said is correct^

Reply:
I'm torn on whether it is easier or harder to read now.  I think we need a third opinion.

Reply:
my first go at it I implemented `def _kubernetes_context_callable()` and `def _katib_config_context_callable()`, but then they had common code, so I moved to this approach you see here. 
another option can be to implement `def _kubernetes_context_callable()` and `def _katib_config_context_callable()`, and another small method that contains the reusable code sth like def `_combine_dicts(dict1, dict2)`,  wdyt?

Reply:
after some discussion, we agreed to implement a function to calculate each context, and to use `dict1.update(dict2)` rather than dict expansion for better readability.